nixos/modules/services/web-apps/alist.nix

```diff
@@ -0,0 +1,294 @@
+                description = ''
+                  Unix socket file path to listen on, default empty, set it to non empty to enable unix socket.
+                '';
+              };
+              unix_file_perm = lib.mkOption {
+                type = lib.types.nullOr lib.types.str;
+                default = null;
+                description = ''
+                  Unix socket file permission, set to the appropriate permissions.
+                '';
+              };
+            };
+            temp_dir = lib.mkOption {
+              type = lib.types.str;
+              default = "/tmp";
+              description = ''
+                The directory to keep temporary files exclusive to alist.
+                In order to prevent AList from generating garbage files when being interrupted,
+                the directory will be cleared every time AList starts, so do not store anything in this directory.
+              '';
+            };
+          };
+        };
+        default = { };
+        description = ''
+          The alist configuration, see https://alist.nn.ci/config/configuration.html for documentation.
+
+          Options containing secret data should be set to an attribute set
+          containing the attribute `_secret` - a string pointing to a file
+          containing the value the option should be set to.
+        '';
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    # Let Alist read tls-related configurations from environment variables
+    services.alist.settings.force = false;
+
+    # Alist may store files in local paths, so make the alist user permanent.
+    users.users.alist = lib.optionalAttrs (cfg.user == null) {
+      description = "Alist user";
+      isSystemUser = true;
+      group = lib.optionalString (cfg.group == null) "alist";
+    };
+    users.groups.alist = lib.optionalAttrs (cfg.group == null) { };
+
+    systemd.services.alist = {
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+      preStart =
+        ''
+          ${utils.genJqSecretsReplacementSnippet cfg.settings "/run/alist/new"}
+        ''
+        + (
+          if cfg.allowModify then
+            ''
+              if [ -e "${cfg.stateDir}/config.json" ]; then
+                cp "${cfg.stateDir}/config.json" /run/alist/old
+                ${lib.getExe pkgs.jq} -s '.[0] * .[1]' /run/alist/old /run/alist/new > /run/alist/result
+                cp /run/alist/result "${cfg.stateDir}/config.json"
+              else
+                cp /run/alist/new "${cfg.stateDir}/config.json"
+              fi
+              rm -f /run/alist/{old,new,result}
+            ''
+          else
+            ''
+              cp /run/alist/new "${cfg.stateDir}/config.json"
+              rm -f /run/alist/new
+            ''
+        );
+      script = ''
+        ${lib.optionalString (cfg.settings.scheme.cert_file != null) ''
+          export ALIST_CERT_FILE="''${CREDENTIALS_DIRECTORY}/tls-cert"
```

Comment:
Could these be placed in `environment` instead?

Reply:
Because of the usage of `CREDENTIALS_DIRECTORY`, is impossible.